src/DynamicData/Cache/Internal/FullJoin.cs

```diff
@@ -118,6 +118,9 @@ internal class FullJoin<TLeft, TLeftKey, TRight, TRightKey, TDestination>(IObser
                     return joinedCache.CaptureChanges();
                 });
 
-                return new CompositeDisposable(leftLoader.Merge(rightLoader).SubscribeSafe(observer), leftCache, rightCache);
+                lock (locker)
```

Comment:
This seems bizarre to me. The resource we need to lock around is the two caches, right? Those aren't actually being accessed here, just passed around by reference. By my reckoning, the only way a read or write occurs on one of the caches, from inside this lock, is if the `.Merge()` stream emits unsynchronized notifications *during* subscription, and that would be a defect inside the `.Merge()` operator.

Reply:
That was my initial thought, but then again this operator by nature produces asynchrony ands the rx guideline is to make sure your custom operators are synchronised.

Reply:
Right, which is why the other recommendation is to rely on operators you use to only emit synchronized notifications.

Basically, this suggests to me that there's a bug inside `.Merge()` which this doesn't actually fix. Unless I'm missing something, at the moment I'm only looking at all this on my phone.

Reply:
I agree, but what is certain is this change fixes the issue.

Reply:
I played around with it last night, and yeah, there really is only two effective fixes, without significantly redesigning the operator: this, or swapping the `.AsObservableCache(false)` calls to `.AsObservableCache(true)`. The issue isn't that notifications aren't being synchronized, it's that there's code running _before_ the first notification, to build up the initial changeset, which accesses the caches shared by all subscribers, and because it's buried inside `.Connect().Subscribe()` we can't lock around it directly.

Reply:
Yeah, originally it used the locked version of the cache but I optimised later but changing that to use unlocked with more operator based locking.